projects/03-captions-api/requirements.md

```diff
@@ -0,0 +1,205 @@
+
+If the request contains the required parameters and their name is correct but their value is invalid (eg: sending an empty string in `url`) the server response should be `422`. More details about 422 status code [here](https://datatracker.ietf.org/doc/html/rfc4918#section-11.2).
+
+*Error response body*
+
+When the response is `4xx` then the response body should contain an error object:
+
+```json
+{
+  "code": "missing_parameters",
+  "title": "Parameter is missing from the request body",
+  "description" "url parameter is missing from the request body. It is a required parameter and the request cannot be processed without it"
+}
+```
+
+More details about these object propertiers [here](https://jsonapi.org/format/#errors).
+
+## Simple Captions
+
+### POST /captions
+
+Request: 
+
+```
+POST /captions
+```
+with the following body: 
+
+```json
+{
+  "caption": {
+    "url": "http://image.url",
+    "text": "caption text"
+  }
+}
+```
+and it will get back the image with the caption text on it.
+
+The response will be: 
+```json 
+{
+  "caption": {
+    "id": 123,
+    "url": "http://image.url/meme.jpg",
+    "text": "caption text",
+    "caption_url": "http://example.com/images/meme.jpb" 
+  },
+}
+```
+The response status will be: 303 and should contain in `Location` header field the `caption_url`. 
+
+### GET /captions
+
+Will return a list of all generated simple captions.
+
+The responde body will be a JSON with the following structure: 
+
+```json
+{
+  "captions": [
+    {
+      "id": 123,
+      "url": "",
+      "text": "",
+      "caption_url": "" 
+    },
+    {
+      "id": 124,
+      "url": "",
+      "text": "",
+      "caption_url: ""
+    }
+  ]
+}
+```
+
+where: 
+- `id` is the generated unique ID
+- `url` is the original URL of the image submitted by the user
+- `text` is the text that should be added on the image
+- `caption_url` is the URL of the generated image.
+
+### GET /captions/:id
+Will return the attributes for the caption with id `:id`
+```json 
+{
+  "caption": {
+    "id": 123,
+    "url": "http://image.url/meme.jpg",
+    "text": "caption text",
+    "caption_url": "http://example.com/images/meme.jpb" 
+  },
+}
+```
+
+### DELETE /captions/:id
+Will delete the caption with that ID and its associated files. 
+The response will be `:200` if the delete was successful. 
+
+## Instagram Captions
+
+### POST /captions/instagrams
+
+API client will do a request: 
+
+```
+POST /captions/instagram
+```
+with the following body: 
+
+```json
+{
+  "image": {
+    "type": "image",
+    "url": "http://image.url",
+    "text": "caption text",
+    "filter": "blackwhite",
+  }
+}
+```
+The response will the same as for the `POST /captions`.
+
+See details about images sizes [here](https://help.instagram.com/1631821640426723), you should crop the image accordingly with the image sizes. 
```

Comment:
Whill this endpoint have to do image processing ? Looks like the image does not have a "caption_url"

Reply:
The `caption_url` is the processed image, the result. I should maybe add an example also here